src/dbnode/storage/shard.go

```diff
@@ -2721,41 +2721,19 @@ func (s *dbShard) AggregateTiles(
 		processedBlockCount atomic.Int64
 		multiErr            xerrors.MultiError
 		dp                  ts.Datapoint
-		id, prevID          ident.ID
 		downsampledValues   = make([]tile.DownsampledValue, 0, 4)
-		tags                ident.TagIterator
-		justStarted         = true
-		lastError           error
 	)
 
 	for readerIter.Next() {
-		seriesIter := readerIter.Current()
+		seriesIter, id, tags := readerIter.Current()
 		prevFrameLastValue := math.NaN()
 		for seriesIter.Next() {
 			frame := seriesIter.Current()
-			id = frame.ID()
-			if justStarted {
-				prevID = id
-				justStarted = false
-				encoder.Reset(opts.Start, 0, targetSchemaDesc)
-			} else if !id.Equal(prevID) {
-				if err := writer.Write(ctx, encoder, prevID, tags); err != nil {
-					s.metrics.largeTilesWriteErrors.Inc(1)
-					lastError = err
-				} else {
-					s.metrics.largeTilesWrites.Inc(1)
-				}
-
-				encoder.Reset(opts.Start, 0, targetSchemaDesc)
-				prevID = id
-			}
-
-			tags = frame.Tags()
+			encoder.Reset(opts.Start, 0, targetSchemaDesc)
```

Comment:
Ah I thought the motivation here was to write the downsampled point from each frame

Reply:
I believe the `encoder` accumulates the data points for the whole series and then gets them persisted, but not 100% sure. Hope @gediminasgu can confirm this.
There's another integration test in src/dbnode/integration/large_tiles_test.go that I commented out during all this debugging hell, try uncommenting it, I hope it would catch the data mismatch.

Reply:
Also @gediminasgu I think it would be useful (and we briefly talked some time ago) if the `write_large_tiles.go` would track the previously written id and verify that `id > previousId`. This should detect the case like here.
Also, makes me wonder who is finalizing the id now. Should be handled identically to the tags, I believe.